### PR TITLE
Adds support for subDomain output

### PR DESCRIPTION
**Pull request by mmcgurn**
This PR adds support for subDomain io and arbitrary labeling using a math function.  This moves us closer to #139 and #46

### CODE REVIEW

File: ablateLibrary/domain/modifiers/distributeWithGhostCells.cpp [hunk 1]
```diff
@@ -0,0 +1,19 @@
+#include "distributeWithGhostCells.hpp"
+#include <utilities/petscError.hpp>
+
+ablate::domain::modifier::DistributeWithGhostCells::DistributeWithGhostCells(int ghostCellDepthIn) : ghostCellDepth(ghostCellDepthIn < 1 ? 1 : ghostCellDepthIn) {}
+void ablate::domain::modifier::DistributeWithGhostCells::Modify(DM &dm) {
+    // Make sure that the flow is setup distributed
+    DM dmDist;
+    // create any ghost cells that are needed
+    DMSetBasicAdjacency(dm, PETSC_TRUE, PETSC_FALSE) >> checkError;
+    DMPlexDistribute(dm, ghostCellDepth, NULL, &dmDist) >> checkError;
```
Comment (knepley) [inline ablateLibrary/domain/modifiers/distributeWithGhostCells.cpp:10]: Note that this is one option for having overlap, but another is to just add the call to `DMPlexDistributeOverlap()` after a regular distribution.

File: ablateLibrary/domain/subDomain.cpp [hunk 2]
```diff
@@ -153,3 +220,158 @@ void ablate::domain::SubDomain::ProjectFieldFunctions(const std::vector<std::sha
...
+
+    // March over the global section
+    PetscSection section;
+    DMGetGlobalSection(sDM, &section) >> checkError;
+
+    PetscInt pStart, pEnd;
+    PetscSectionGetChart(section, &pStart, &pEnd) >> checkError;
+
+    // For each field in the subDM
+    for (std::size_t i = 0; i < subFields.size(); i++) {
```
Comment (knepley) [inline ablateLibrary/domain/subDomain.cpp:343]: Note that this mostly reproduces the code in `DMCreateSubDM()` which returns an IS with these indices.

File: ablateLibrary/domain/subDomain.cpp [hunk 3]
```diff
@@ -137,7 +166,45 @@ PetscObject ablate::domain::SubDomain::GetPetscFieldObject(const Field& field) {
...
+
+    if (label == nullptr) {
+        DMProjectFunction(dm, time, &fieldFunctions[0], &fieldContexts[0], INSERT_VALUES, globVec) >> checkError;
+
+    } else {
+        DMProjectFunctionLabel(dm, time, label, region->GetValues().size(), &region->GetValues()[0], -1, NULL, &fieldFunctions[0], &fieldContexts[0], INSERT_VALUES, globVec);
+    }
+}
+
+void ablate::domain::SubDomain::ProjectFieldFunctionsToSubDM(const std::vector<std::shared_ptr<mathFunctions::FieldFunction>>& initialization, Vec globVec, PetscReal time) {
```
Comment (knepley) [inline ablateLibrary/domain/subDomain.cpp:196]: Note that the `dm` input to Project can have different fields than the DM attached to the input vector, so you can compress out fields with the single call if you want.

**Review by knepley — APPROVED** [review]
I made some comments

File: ablateLibrary/domain/subDomain.cpp (hunk 3 above)
Comment (mmcgurn) [reply ablateLibrary/domain/subDomain.cpp:196]: Thanks!  Does that mean that the fieldFunctions should be sized/correspond to vec and not the dm?

File: ablateLibrary/domain/subDomain.cpp (hunk 2 above)
Comment (mmcgurn) [reply ablateLibrary/domain/subDomain.cpp:343]: Thanks @knepley!  Are there any support functions for moving data with the IS or should I iterate through it accessing array data?

**Comment by mmcgurn** [comment]
> I made some comments

Thanks!  I think I'm going to merge this in for now and go back and cleanup and use DMCreateSubDM in the future.

File: ablateLibrary/domain/subDomain.cpp (hunk 2 above)
Comment (knepley) [reply ablateLibrary/domain/subDomain.cpp:343]: Yes, you call [VecISCopy](https://petsc.org/main/docs/manualpages/Vec/VecISCopy.html), (VecISSet)[https://petsc.org/main/docs/manualpages/Vec/VecISSet.html], [VecISAXPY](https://petsc.org/main/docs/manualpages/Vec/VecISAXPY.html)

File: ablateLibrary/domain/subDomain.cpp (hunk 3 above)
Comment (knepley) [reply ablateLibrary/domain/subDomain.cpp:196]: I do not completely understand the question, but we could go over it at the meeting.

**Comment by knepley** [comment]
Sounds good.

**Pull request merged**